--- elements/wifi/bk.radiotapdecap.cc	2009-08-25 18:09:14.000000000 +0100
+++ elements/wifi/radiotapdecap.cc	2009-08-25 20:49:47.000000000 +0100
@@ -122,8 +122,19 @@
 	struct ieee80211_radiotap_header *th = (struct ieee80211_radiotap_header *) p->data();
 	struct click_wifi_extra *ceh = WIFI_EXTRA_ANNO(p);
 	if (rt_check_header(th, p->length())) {
+		memset((void*)ceh, 0, sizeof(struct click_wifi_extra));
 		ceh->magic = WIFI_EXTRA_MAGIC;
 
+		if (rt_el_present(th, IEEE80211_RADIOTAP_FLAGS)) {
+			u_int8_t flags = *((u_int8_t *) rt_el_offset(th, IEEE80211_RADIOTAP_FLAGS));
+			if (flags & IEEE80211_RADIOTAP_F_DATAPAD) {
+			     ceh->pad = 1;
+			}
+			if (flags & IEEE80211_RADIOTAP_F_FCS) {
+			     p->take(4);
+			}
+		}
+
 		if (rt_el_present(th, IEEE80211_RADIOTAP_RATE)) {
 			ceh->rate = *((u_int8_t *) rt_el_offset(th, IEEE80211_RADIOTAP_RATE));
 		}
@@ -141,9 +152,16 @@
 			ceh->silence = *((u_int8_t *) rt_el_offset(th, IEEE80211_RADIOTAP_DB_ANTNOISE));
 
 		if (rt_el_present(th, IEEE80211_RADIOTAP_RX_FLAGS)) {
+			click_chatter("RX Flags Present");
 			u_int16_t flags = le16_to_cpu(*((u_int16_t *) rt_el_offset(th, IEEE80211_RADIOTAP_RX_FLAGS)));
-			if (flags & IEEE80211_RADIOTAP_F_RX_BADFCS)
+			if (flags & IEEE80211_RADIOTAP_F_RX_BADFCS) {
+				click_chatter("CRC FAIL!");
 				ceh->flags |= WIFI_EXTRA_RX_ERR;
+			} else {
+				click_chatter("Good CRC..");
+			}
+		} else {
+			click_chatter("No RX Flags :(");
 		}
 
 		if (rt_el_present(th, IEEE80211_RADIOTAP_TX_FLAGS)) {
@@ -151,16 +169,13 @@
 			ceh->flags |= WIFI_EXTRA_TX;
 			if (flags & IEEE80211_RADIOTAP_F_TX_FAIL)
 				ceh->flags |= WIFI_EXTRA_TX_FAIL;
-
-			if (flags & IEEE80211_RADIOTAP_F_FCS) {
-				p->take(4);
-			}
 		}
 
 		if (rt_el_present(th, IEEE80211_RADIOTAP_DATA_RETRIES))
 			ceh->retries = *((u_int8_t *) rt_el_offset(th, IEEE80211_RADIOTAP_DATA_RETRIES));
 
 		p->pull(th->it_len);
+		p->set_mac_header(p->data());  // reset mac-header pointer
 	}
 
   return p;
